Remove lines mandating region be specified by KMS keyring generator (#50)
awslabs/aws-encryption-sdk-specification: framework/kms-keyring.md

### Client Supplier

> A function that maps a region string to a KMS client that can make the following API calls in the given AWS region:

Reviewer: I think we want to be clear what the inputs/outputs are. The input is a string, the output is a KMS client.

Author: Can you help me understand in what way is this sentence unclear about that?

Reviewer: The latest wording addresses my concern. I was just pointing out that using `maps` isn't as clear as explicitly stating what the input is and what the output is.

### Client Supplier

> A function which MUST take as input either a region string (e.g. `us-east-1`), or some value denoting an unknown region,

Reviewer 2: ```suggestion
A function that MUST take as input either a region string (e.g. `us-east-1`), or some value denoting an unknown region,
```

> and MAY return a KMS Client which can make the following API calls in the given AWS region:

Reviewer 2: ```suggestion
and MAY return a KMS Client that can make the following API calls in the given AWS region:
```

- A CMK [alias](https://docs.aws.amazon.com/kms/latest/developerguide/programming-aliases.html) (e.g. "alias/MyCryptoKey")
- A well-formed key ARN (e.g. arn:aws:kms:us-east-1:999999999999:key/01234567-89ab-cdef-fedc-ba9876543210)

> - A well-formed alias ARN (e.g. arn:aws:kms:us-east-1:999999999999:alias/MyCryptoKey)

Reviewer 3: From the KMS side, `/` is the _only_ supported version.

and this keyring has a [generator](#generator) defined, and onEncrypt MUST attempt to generate a new plaintext data key
and encrypt that data key by calling [KMS GenerateDataKey](#kms-generatedatakey).

If an AWS region can be extracted from the [generator](#generator), then the [KMS client](#kms-client) that calls
[KMS GenerateDataKey](#kms-generatedatakey) MUST be the client returned by the [client supplier](#client-supplier)
when given that region as input. If an AWS region cannot be extraced from the [generator](#generator),

> then the KMS Keyring MAY choose an arbitrary region.

Reviewer: We might want to clarify what we mean by `May choose an arbitrary region`. What we really mean is that if you can't get the region from the key, then the KMS keyring is allowed to use a KMS client from any arbitrary region to make the call to KMS. Maybe just say "If the AWS region cannot be extracted...the KMS Keyring MAY use a KMS client from any region"

Reviewer 2: This is where I hit a bit of a sticking point: Who is the actor here?

# Region discernible from key ID

1. The KMS Keyring *through some undefined mechanism* attempts to determine region from key ID.
1. The KMS Keyring passes the region name to the client supplier.
1. The client supplier provides the KMS Keyring with a client.
1. The KMS Keyring uses that client.

# Region not discernible from key ID

## Keyring is the actor

1. The KMS Keyring *through some undefined mechanism* attempts to determine region from key ID.
1. The KMS Keyring fails to determine region from key ID.
1. The KMS Keyring *through some undefined mechanism* obtains a client.

## Client Supplier is the actor

1. The KMS Keyring *through some undefined mechanism* attempts to determine region from key ID.
1. The KMS Keyring fails to determine region from key ID.
1. The KMS Keyring tells the Client Supplier that it doesn't know what the region is.
1. The Client Supplier MAY provide the KMS Keyring with a client.
1. The KMS Keyring uses that client or fails if the Client Supplier fails to provide one.

Reviewer: There are two points being brought up here:
1. We should specify that the client supplier is responsible for determining default/"I can't get a region" behavior instead of the keyring. As said above, this makes sense because otherwise we violate the API contract.
2. In the view of a client supplier, it can be a bit tricky what the default behavior SHOULD be. Generally the supplier should make use of the AWS SDK defualt region, however the "limit regions" supplier example above shows how this might not always be the desired behavior. I think more precisely defining the restrictions on client supplier behavior is out of scope for this PR.

As such, I think these are the next steps for this PR:

- Update client supplier interface to say that it MUST support some input (i.e. empty string) to indicate that the region in unknown
- Update L66 to say that if the region can't be determined, then the Keyring MUST input the "I don't know" input to the client supplier, and if the client supplier returns a client the Keyring MUST use that client.

@mattsb42-aws @seebees Thoughts?

Reviewer 3: Good to me.

Reviewer 2: Sounds good!

Author: As I read the spec, the Client Supplier is an interface which we let the customer implement. As such, I don't think it makes sense to define default behavior.

Author: How about:

> ### Client Supplier
> A function which MUST take as input either a region string (e.g. `us-east-1`), or some value denoting an unknown region, and MAY return a KMS Client which can make the following API calls in the given AWS region:

and

> If an AWS region cannot be extracted from the generator then the KMS Keyring MUST input a value denoting an unknown region. If the client supplier does not provide any client for the given region for this Encrypt call, OnEncrypt MUST skip that particular CMK.

I've opted not to define the value indicating an unknown region, as that seems fairly language specific.

Reviewer 2: `s/which/that/` but otherwise LGTM @MatthewBennington

Reviewer: new wording LGTM

and this keyring has a [generator](#generator) defined, and onEncrypt MUST attempt to generate a new plaintext data key
and encrypt that data key by calling [KMS GenerateDataKey](#kms-generatedatakey).

> If an AWS region can be extracted from the [generator](#generator), then the [KMS client](#kms-client) that calls

Reviewer 3: This is also true of the KMS Encrypt call below.

[KMS GenerateDataKey](#kms-generatedatakey) MUST be the client returned by the [client supplier](#client-supplier)

> when given that region as input.

Reviewer 2: This feels like it is drawing an odd relationship that I don't think exists. The way this is worded make it sound to me like we are saying that there is something special about the region name extracted from *this* ARN that makes it different from the same region name extracted from *some other* ARN. This is not the case (though, side note, this is why I wanted to the client provider to accept a key name rather than a region name, but that's a different conversation).

I think we can simplify this to something along the lines of the following.

**Client Provider**

* The client provider MUST require a region name.
* The client provider MUST either provide a client for the requested region or fail.

**Keyring**

* If the key ID is an ARN then the keyring MUST call the client provider with the region in the ARN.
* If the key ID is *not* an ARN then the keyring MAY select a region with which to call the client provider.

Author: > The way this is worded make it sound to me like we are saying that there is something special about the region name extracted from this ARN that makes it different from the same region name extracted from some other ARN.

As I understand the spec, there is something special about this region. No other region derived from an ARN determines the region used by the client (or after this PR, the region given to the client supplier).

Reviewer 2: There is something special about the generator key. That generator key happens to exist in a single region. The region is a side effect.

Author: How is this?

> If the generator is formatted as either a well-formed key ARN, or a well-formed alias ARN, then an AWS region MUST be derived from the generator and used to call the client supplier. The client produced as output of the client supplier MUST be used to call KMS Encrypt/Decrypt/GenerateDataKey.

Reviewer 3: >The client provider MUST require a region name.

I disagree.
The AWS SDK can pick up a default region in the same way it gets credentials.
The client provider is the interface to the AWS SDK.
Let the SDK handle it.
If there is any additional complication regarding what a default region should be,
putting this logic in a function is much easier for customers.

I have an example of what I'm thinking about here: https://github.com/aws/aws-encryption-sdk-javascript/pull/231

and this keyring has a [generator](#generator) defined, and onEncrypt MUST attempt to generate a new plaintext data key
and encrypt that data key by calling [KMS GenerateDataKey](#kms-generatedatakey).

If an AWS region can be extracted from the [generator](#generator), then the [KMS client](#kms-client) that calls
[KMS GenerateDataKey](#kms-generatedatakey) MUST be the client returned by the [client supplier](#client-supplier)
when given that region as input.
If an AWS region cannot be extracted from the [generator](#generator) then the KMS Keyring MUST input a value denoting an unknown region.

> If the [client supplier](#client-supplier) does not provide any client for the given region for this [KMS GenerateDataKey](#kms-generatedatakey) call, OnEncrypt MUST not modify the [encryption materials](#structures.md#encryption-materials.md)

Reviewer 2: ```suggestion
If the [client supplier](#client-supplier) does not provide any client for the given region for this [KMS GenerateDataKey](#kms-generatedatakey) call, OnEncrypt MUST NOT modify the [encryption materials](#structures.md#encryption-materials.md)
```

modify the [encryption materials](#structures.md#encryption-materials) and MUST fail.

If the call succeeds, OnEncrypt MUST verify that the response `Plaintext` length matches the specification

> of the [algorithm suite](algorithm-suites.md)'s Key Derivation Input Length field. If it does not, OnEncrypt MUST fail.

Reviewer: Here and elsewhere: newline after `.`

- append a new [record](#structures.md#record) to the [keyring trace](#structures.md#keyring-trace)
  in the input [encryption materials](#structures.md#encryption-materials), constructed as follows:
  - The string field KeyNamespace contains "aws-kms".
  - The string field KeyName contains the value of the KMS Keyring's [generator](#generator) field.
  - The [flags](#structures.md$flags) field of this record includes exactly the following flags:
    - [GENERATED DATA KEY](#structures.md#supported-flags)
    - [ENCRYPTED DATA KEY](#structures.md#supported-flags)
    - [SIGNED ENCRYPTION CONTEXT](#structures.md#supported-flags)

> 

Reviewer: Without this qualifier is it clear what `plaintext data key` is referring to on the next line? Maybe there is a way to restate this which is less confusing, or do you think it's not necessary?

Author: Good point about clarifying `plaintext data key`. I removed this line because I think it creates the appearance of an else to the if on line 154. I will try to find wording that addresses this.

- append a new [record](#structures.md#record) to the [keyring trace](#structures.md#keyring-trace)
  in the input [encryption materials](#structures.md#encryption-materials), constructed as follows:
  - The string field KeyNamespace contains "aws-kms".
  - The string field KeyName contains the value of the KMS Keyring's [generator](#generator) field.
  - The [flags](#structures.md$flags) field of this record includes exactly the following flags:
    - [GENERATED DATA KEY](#structures.md#supported-flags)
    - [ENCRYPTED DATA KEY](#structures.md#supported-flags)
    - [SIGNED ENCRYPTION CONTEXT](#structures.md#supported-flags)

>     

Reviewer 2: ```suggestion

```

To attempt to encrypt the plaintext data key with a particular CMK, OnEncrypt MUST call [KMS Encrypt](#kms-encrypt).

> For each [KMS Encrypt](#kms-encrypt) call, if an AWS region can be extracted from the CMK ARN, then the

Reviewer 3: ```suggestion
For each [KMS Encrypt](#kms-encrypt) call, if an AWS region can be extracted from the [Key Id](#key-ids), then the
```
Part of what we are doing here is dealing with things that are not ARNs

> For each [KMS Encrypt](#kms-encrypt) call, if an AWS region can be extracted from the [Key Id](#key-ids), then the

Reviewer 2: ```suggestion
For each [KMS Encrypt](#kms-encrypt) call, if an AWS region can be extracted from the [Key ID](#key-ids), then the
```

[KMS client](#kms-client) that calls [KMS Encrypt](#kms-encrypt) MUST be the client returned by the
[client supplier](#client-supplier) when given that region as input.

> If an AWS region cannot be extracted from the CMK ARN then the KMS Keyring MUST input a value denoting an unknown region.

Reviewer 2: ```suggestion
If an AWS region cannot be extracted from the Key ID then the KMS Keyring MUST input a value denoting an unknown region.
```

Author: "CMK ARN" is more consistent with the rest of the document. "Key ID" elsewhere is only used to refer elements of the "Key IDs" list.

In my opinion we should revise the spec to have better wording around "CMK" vs "CMK ARN" vs "Key ID". But that is not within the scope of this PR.

Author: Actually, re-reading through the file, I see the "Key ID" is more consistent within the paragraph. I will add this change.

- append a new [record](#structures.md#record) to the [keyring trace](#structures.md#keyring-trace)
  in the input [encryption materials](#structures.md#encryption-materials), constructed as follows:
  - The string field KeyNamespace contains "aws-kms".

>   - The string field KeyName contains the CMK ARN used for [KMS Encrypt](#kms-encrypt).

Reviewer 3: Both the C and JS ESDKs do not do this.  They both use the value returned from KMS.

To attempt to decrypt a particular [encrypted data key](#structures.md#encrypted-data-key),
OnDecrypt MUST call [KMS Decrypt](#kms-decrypt).

> If an AWS region can be extracted from the [generator](#generator), then the [KMS client](#kms-client) that calls

Reviewer 3: On the other hand.  For decrypt the CMK _must_ be the ARN.

Author: The CMK used for decryption must be an ARN, but that CMK is not necessarily the generator. I don't understand how this comment relates to this line, can you clarify?

Reviewer: For decryption, only keys with a well formed key ARN will produce a successful decryption. This is because the keyring will only attempt to call KMS Decrypt for EDKs where the keyProvider matches a keyID. Because the EDK is serialized to always use the key ARN as the keyProvider, the only way to get the KMS Keyring to call KMS Decrypt for an EDK is if you include the key ARN in the keyIDs/generator.

Thus, if we ever actually get to attempting to call KMS, we know that the key in question has a region. HOWEVER, the generator could still just be an alias... I think it still makes sense to specify the "everything is a well formed ARN" case, and leave specifying other behavior as part of #49

Author: > Because the EDK is serialized to always use the key ARN as the keyProvider, the only way to get the KMS Keyring to call KMS Decrypt for an EDK is if you include the key ARN in the keyIDs/generator.

If I understand correctly, then if a given EDK's KeyProvider matches one of the keyring's KeyIDs, then it will attempt to decrypt. In this scenario the keyring would need to create a client by inputting a region into the client supplier. To determine that region it would need to parse the generator string, which may be an alias.

To attempt to decrypt a particular [encrypted data key](#structures.md#encrypted-data-key),
OnDecrypt MUST call [KMS Decrypt](#kms-decrypt).

> If an AWS region can be extracted from the [encrypted data key](structures.md#encrypted-data-key)'s

Reviewer: An AWS region MUST be extractable from the EDK. This is because the EDK key provider info serializes the keyARN, and never the alias. If the KMS Keyring cannot extract the region in this case, it should result in an error.

To attempt to decrypt a particular [encrypted data key](#structures.md#encrypted-data-key),
OnDecrypt MUST call [KMS Decrypt](#kms-decrypt).

For each [KMS Decrypt](#kms-decrypt) call, if an AWS region can be extracted from the [encrypted data key](structures.md#encrypted-data-key)'s [key provider info](structures.md#key-provider-information).

> The KMS Keyring MUST call [KMS Decrypt](#kms-decrypt) using the client supplied by the [client supplier](#client-supplier), given the region as input.

Reviewer: Whoops, this wording needs to also be updated to remove the "if an AWS region can be extracted".

To attempt to decrypt a particular [encrypted data key](#structures.md#encrypted-data-key),
OnDecrypt MUST call [KMS Decrypt](#kms-decrypt).

For each [KMS Decrypt](#kms-decrypt) call, if an AWS region can be extracted from the [encrypted data key](structures.md#encrypted-data-key)'s [key provider info](structures.md#key-provider-information).
The KMS Keyring MUST call [KMS Decrypt](#kms-decrypt) using the client supplied by the [client supplier](#client-supplier), given the region as input.

If an AWS region cannot be extracted from the [encrypted data key](structures.md#encrypted-data-key)'s [key provider info](structures.md#key-provider-information) then the KMS Keyring MUST input to the [client supplier](#client-supplier) a value denoting an unknown region.

> If the client supplier does not provide any client for the given region for this Decrypt call, OnDecrypt MUST skip that particular [encrypted data key](#encrypted-data-key).

Reviewer: We shouldn't specify this for OnDecrypt, as the specified behavior above will never allow us to try to obtain a region from a key that doesn't have a region. Instead, the spec at this point should state that you MUST get the region from the key. If for whatever the implementation is unable to get a region from the key at this point, something is really wrong, and the implementation should fail due to being unable to follow the specification.

Author: Good catch

- append a new [record](#structures.md#record) to the [keyring trace](#structures.md#keyring-trace)
  in the input [encryption materials](#structures.md#encryption-materials), constructed as follows:

>   - The string field KeyNamespace contains "aws-kms".

Reviewer 2: ```suggestion
  - The string field KeyNamespace is "aws-kms".
```

>   - The string field KeyName contains the [encrypted data key's key provider info](structures.md#key-provider-information).

Reviewer 2: ```suggestion
  - The string field KeyName is the [encrypted data key's key provider info](structures.md#key-provider-information).
```
